### PR TITLE
[gosrc2cpg] - Generic parameter type handling

### PR DESCRIPTION
**Pull request by ankit-privado**
Added support to handle the ast creation for single generic type method.

### CODE REVIEW

**Review by pandurangpatil — CHANGES_REQUESTED** [review]

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala [hunk 1]
```diff
@@ -32,6 +32,7 @@ class AstCreator(val relPathFileName: String, val parserResult: ParserResult)(im
 
   protected val methodAstParentStack: Stack[NewNode]                 = new Stack()
   protected val scope: Scope[String, (NewNode, String), NewNode]     = new Scope()
+  protected val genericTypeMapping: mutable.Map[String, String]      = mutable.Map.empty
```
[the remark this answers is not in the document]
Comment (ankit-privado) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala:35]: yes for this draft PR only i have used `map`, will change it to `scope`.

File: joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/passes/ast/AstCreationForGenericTests.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      val cpg = code("""
+          |package main
+          |func foo[T float](value T) {}
+          |""".stripMargin)
+
+      val List(param) = cpg.method.name("foo").parameter.name("value").l
+      param.typeFullName shouldBe "float"
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (ankit-privado) [reply joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/passes/ast/AstCreationForGenericTests.scala:41]: yes sure will add.

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala [hunk 3]
```diff
@@ -36,9 +36,12 @@ trait AstCreatorHelper { this: AstCreator =>
         parserNodeCache.addOne(json(ParserKeys.NodeId).num.toLong, pni)
         node match
           case CallExpr =>
-            json(ParserKeys.Fun)(ParserKeys.Obj).objOpt.map(obj => {
```
Comment (pandurangpatil) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala:39]: You can revert these changes and take the latest updates from the master. A recent PR merge fixes this one as well as few more use cases.

Comment (ankit-privado) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala:39]: okay sure.

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (ankit-privado) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala]: used map as discussed.

File: joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/passes/ast/AstCreationForGenericTests.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (ankit-privado) [reply joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/passes/ast/AstCreationForGenericTests.scala:106]: done.

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala (hunk 3 above)
Comment (ankit-privado) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala]: done.

**Review by DavidBakerEffendi — CHANGES_REQUESTED** [review]

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala [hunk 4]
```diff
@@ -109,7 +109,11 @@ trait AstCreatorHelper { this: AstCreator =>
       .toMap
   }
 
-  private def generateTypeFullName(typeName: String = "", aliasName: Option[String] = None): String = {
+  private def generateTypeFullName(
+    typeName: String = "",
```
[the remark this answers is not in the document]
Comment (ankit-privado) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala:113]: It's been there from the beginning, okay i will change it.

[the remark this answers is not in the document]
Comment (DavidBakerEffendi) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala:113]: @ankit-privado I see, if it's linked to too many things then we can skip

[the remark this answers is not in the document]
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala:113]: @DavidBakerEffendi I will handle that in separate PR.

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala [hunk 5]
```diff
@@ -121,16 +125,23 @@ trait AstCreatorHelper { this: AstCreator =>
             // NOTE: If the given type is not found in primitiveTypeMap.
             // Then we are assuming the type is custom type defined inside same pacakge as that of current file's package.
             // This assumption will be invalid when another package is imported with alias "."
-            Defines.primitiveTypeMap.getOrElse(typeName, s"${fullyQualifiedPackage}.${typeName}")
+            if (genericTypeMethodMap.contains(typeName)) {
+              genericTypeMethodMap(typeName).mkString("|")
+            } else {
+              Defines.primitiveTypeMap.getOrElse(typeName, s"${fullyQualifiedPackage}.${typeName}")
```
[the remark this answers is not in the document]
Comment (ankit-privado) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala:131]: done.

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (ankit-privado) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreatorHelper.scala:113]: yes sure.

**Review by DavidBakerEffendi — APPROVED** [review]
